CustomerData/CustomerTagging.php

```diff
@@ -76,10 +80,36 @@ public function getSectionData()
                 'first_name' => $customer->getFirstname(),
                 'last_name' => $customer->getLastname(),
                 'email' => $customer->getEmail(),
-                'hcid' => NostoHelperData::generateVisitorChecksum($nostoCustomerId),
+                'hcid' => $this->generateVisitorChecksum($nostoCustomerId),
+                'customer_reference' => $this->generateCustomerReference($customer)
             ];
         }
 
         return $data;
     }
+
+    /**
+     * Return the checksum for for the customer tagging i.e hashed cookie identifier or HCID for
+     * short. This is used to sign the tagging so that if it is in fact cached, the cookie and
+     * tagging signature won't match and we'll be able to show a warning.
+     *
+     * @param string $string
+     * @return string
+     */
+    public static function generateVisitorChecksum($string)
+    {
+        return hash(self::VISITOR_HASH_ALGO, $string);
+    }
+
+    /**
+     * Return the checksum / customer reference for customer
+     *
+     * @param CustomerInterface $customer
+     * @return string
+     */
+    public function generateCustomerReference(CustomerInterface $customer)
+    {
+        /** @noinspection PhpUndefinedMethodInspection */
+        return uniqid(substr(md5($customer->getId() . $customer->getEmail()), 0, 8), true);
```

Comment:
Why use md5? I would use sha256 here as well, unless the performance impact is bad.
What does uniqid do, does it guarantee uniqueness?